File: client/field.go
package client

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// Field helpers are thin sugar over Publish + GetState. Nothing here uses
// server routes that don't already exist — a field mutation is just a
// publish and a value fetch is just a get_state.

type StringField struct {
	c     *Client
	Topic string
}

func (c *Client) StringField(topic string) *StringField { return &StringField{c: c, Topic: topic} }

func (f *StringField) Set(ctx context.Context, value string) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "str_set",
		Payload: mustPayload(map[string]any{"value": value})})
}

func (f *StringField) Delete(ctx context.Context) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "str_delete"})
}

// Get returns (value, exists, error). exists=false is not an error — it just
// means the field has never been set (or was deleted).
func (f *StringField) Get(ctx context.Context) (string, bool, error) {
	raw, err := f.c.GetState(ctx, f.Topic)
	if err != nil {
		return "", false, err
	}
	if len(raw) == 0 {
		return "", false, nil
	}
	var s struct {
		Value  string `json:"value"`
		Exists bool   `json:"exists"`
	}
	if err := json.Unmarshal(raw, &s); err != nil {
		return "", false, err
	}
	return s.Value, s.Exists, nil
}

type IntField struct {
	c     *Client
	Topic string
}

func (c *Client) IntField(topic string) *IntField { return &IntField{c: c, Topic: topic} }

func (f *IntField) Set(ctx context.Context, value int64) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "int_set",
		Payload: mustPayload(map[string]any{"value": value})})
}

func (f *IntField) Incr(ctx context.Context, delta int64) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "int_incr",
		Payload: mustPayload(map[string]any{"delta": delta})})
}

func (f *IntField) Decr(ctx context.Context, delta int64) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "int_decr",
		Payload: mustPayload(map[string]any{"delta": delta})})
}

func (f *IntField) Delete(ctx context.Context) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "int_delete"})
}

func (f *IntField) Get(ctx context.Context) (int64, bool, error) {
	raw, err := f.c.GetState(ctx, f.Topic)
	if err != nil {
		return 0, false, err
	}
	if len(raw) == 0 {
		return 0, false, nil
	}
	var s struct {
		Value  int64 `json:"value"`
		Exists bool  `json:"exists"`
	}
	if err := json.Unmarshal(raw, &s); err != nil {
		return 0, false, err
	}
	return s.Value, s.Exists, nil
}

type TimeField struct {
	c     *Client
	Topic string
}

func (c *Client) TimeField(topic string) *TimeField { return &TimeField{c: c, Topic: topic} }

func (f *TimeField) Set(ctx context.Context, t time.Time) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "time_set",
		Payload: mustPayload(map[string]any{"value": t.UTC().Format(time.RFC3339Nano)})})
}

// Now records the server's current time (server-side, ignores clock skew).
func (f *TimeField) Now(ctx context.Context) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "time_now"})
}

func (f *TimeField) Add(ctx context.Context, d time.Duration) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "time_add",
		Payload: mustPayload(map[string]any{"seconds": int64(d / time.Second)})})
}

func (f *TimeField) Delete(ctx context.Context) (uint64, error) {
	return f.c.Publish(ctx, &Event{Topic: f.Topic, Type: "time_delete"})
}

func (f *TimeField) Get(ctx context.Context) (time.Time, bool, error) {
	raw, err := f.c.GetState(ctx, f.Topic)
	if err != nil {
		return time.Time{}, false, err
	}
	if len(raw) == 0 {
		return time.Time{}, false, nil
	}
	var s struct {
		Value  time.Time `json:"value"`
		Exists bool      `json:"exists"`
	}
	if err := json.Unmarshal(raw, &s); err != nil {
		return time.Time{}, false, err
	}
	return s.Value, s.Exists, nil
}

func mustPayload(v any) json.RawMessage {
	b, err := json.Marshal(v)
	if err != nil {
		panic(fmt.Errorf("field payload marshal: %w", err))
	}
	return b
}

File: clients/rust/src/field.rs
//! Typed helpers over Publish + GetState for scalar-field topics.
//!
//! Nothing here is fundamentally different from calling `Client::publish` with
//! the right event type — the helpers just fix up the type strings and
//! payload shapes so callers can't get them wrong.

use crate::Client;
use serde_json::json;

pub type BoxError = Box<dyn std::error::Error + Send + Sync>;

pub struct StringField {
    c: Client,
    pub topic: String,
}

impl StringField {
    pub(crate) fn new(c: Client, topic: String) -> Self { Self { c, topic } }

    pub async fn set(&self, value: &str) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "str_set", Some(json!({"value": value}))).await
    }
    pub async fn delete(&self) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "str_delete", None).await
    }
    pub async fn get(&self) -> Result<(String, bool), BoxError> {
        let s = self.c.get_state(&self.topic).await?;
        match s {
            None => Ok((String::new(), false)),
            Some(v) => Ok((
                v.get("value").and_then(|x| x.as_str()).unwrap_or("").to_string(),
                v.get("exists").and_then(|x| x.as_bool()).unwrap_or(false),
            )),
        }
    }
}

pub struct IntField {
    c: Client,
    pub topic: String,
}

impl IntField {
    pub(crate) fn new(c: Client, topic: String) -> Self { Self { c, topic } }

    pub async fn set(&self, value: i64) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "int_set", Some(json!({"value": value}))).await
    }
    pub async fn incr(&self, delta: i64) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "int_incr", Some(json!({"delta": delta}))).await
    }
    pub async fn decr(&self, delta: i64) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "int_decr", Some(json!({"delta": delta}))).await
    }
    pub async fn delete(&self) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "int_delete", None).await
    }
    pub async fn get(&self) -> Result<(i64, bool), BoxError> {
        let s = self.c.get_state(&self.topic).await?;
        match s {
            None => Ok((0, false)),
            Some(v) => Ok((
                v.get("value").and_then(|x| x.as_i64()).unwrap_or(0),
                v.get("exists").and_then(|x| x.as_bool()).unwrap_or(false),
            )),
        }
    }
}

pub struct TimeField {
    c: Client,
    pub topic: String,
}

impl TimeField {
    pub(crate) fn new(c: Client, topic: String) -> Self { Self { c, topic } }

    /// value must be RFC3339 (e.g. "2026-01-01T00:00:00Z").
    pub async fn set_rfc3339(&self, value: &str) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "time_set", Some(json!({"value": value}))).await
    }
    pub async fn now(&self) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "time_now", None).await
    }
    pub async fn add(&self, seconds: i64) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "time_add", Some(json!({"seconds": seconds}))).await
    }
    pub async fn delete(&self) -> Result<u64, BoxError> {
        self.c.publish(&self.topic, "time_delete", None).await
    }
    /// Returns (rfc3339-string, exists). Parsing to chrono is left to callers.
    pub async fn get(&self) -> Result<(String, bool), BoxError> {
        let s = self.c.get_state(&self.topic).await?;
        match s {
            None => Ok((String::new(), false)),
            Some(v) => Ok((
                v.get("value").and_then(|x| x.as_str()).unwrap_or("").to_string(),
                v.get("exists").and_then(|x| x.as_bool()).unwrap_or(false),
            )),
        }
    }
}
